cytotable/convert.py

```diff
@@ -245,34 +248,41 @@ def _get_table_chunk_offsets(
 
             return None
 
-    # find chunk offsets from sql statement
     elif sql_stmt is not None:
-        # gather the total rowcount from csv or sqlite data input sources
         with _duckdb_reader() as ddb_reader:
-            rowcount = int(
-                ddb_reader.execute(
-                    # nosec
-                    f"SELECT COUNT(*) FROM ({sql_stmt})"
-                ).fetchone()[0]
-            )
+            sql_query = f"SELECT {page_key} FROM ({sql_stmt}) ORDER BY {page_key}"
+            keys = ddb_reader.execute(sql_query).fetchall()
+            keys = [key[0] for key in keys]
+
+    # Create chunks of keys
+    chunks = []
+    i = 0
+    while i < len(keys):
+        start_key = keys[i]
+        end_index = min(i + chunk_size, len(keys)) - 1
+        end_key = keys[end_index]
+
+        # Ensure non-overlapping by incrementing the start of the next range
+        next_start_index = end_index + 1
+        if next_start_index < len(keys):
+            next_start_key = keys[next_start_index]
+            while next_start_key == end_key and next_start_index + 1 < len(keys):
+                next_start_index += 1
+                next_start_key = keys[next_start_index]
+            chunks.append((start_key, next_start_key - 1))
+        else:
+            chunks.append((start_key, end_key))
 
-    return list(
-        range(
-            0,
-            # gather rowcount from table and use as maximum for range
-            rowcount,
-            # step through using chunk size
-            chunk_size,
-        )
-    )
+        i = next_start_index
```

Comment:
I think this code is fine as-is, but for brevity you might consider using the `grouper()` function as described in the [itertools docs](https://docs.python.org/3.10/library/itertools.html) -- unfortunately I can't link to the function directly, but you'll find it if you search for "def grouper".

A simplified version of that function, [`chunked()`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.chunked), is included in the `more-itertools` library, but I don't know if you want to take an extra dependency for such a simple function.

You could use the `chunked()` function like so to produce your set of non-overlapping intervals:
```python
chunks = [(x[0], x[-1]) for x in chunked(keys, chunk_size)]
```

If you instead use `grouper()`, you'll have to account for the case in which a group ends up padded with `None` if it can't construct a group of size `chunk_size`.
```python
[(x[0], x[-1] if x[-1] is not None else x[0]) for x in grouper(keys, chunk_size)]
```

It seems that SQL's `BETWEEN` is inclusive, so a page like `(100, 100)` will still match the keys with value `100`.

Reply:
Thanks I'll explore this a bit more! I found that `grouper()` was challenging to use with concerns to overlaps that should respect the chunk size. There's likely a more elegant solution as you suggested.

Reply:
Thanks again. I reworked this to simplify and demonstrate reasoning. I decided to create a function out of this work and test it individually to ensure we know what we're receiving (and why the implementation doesn't use existing algorithms in built-in's). My hope here is that the test cases help demonstrate better what output we're expecting (it's a bit different than `grouper`, `groupby`, or `more_itertools.chunked` provide based on repeated value and SQL-specific expectations).